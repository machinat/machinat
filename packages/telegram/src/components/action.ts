import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import makeTelegramComponent from '../utils/makeTelegramComponent.js';
import {
  TelegramSegmentValue,
  UploadingFileInfo,
  TelegramComponent,
} from '../types.js';

/**
 * @category Props
 */
export type ForwardMessageProps = {
  /** Unique identifier for the chat where the original message was sent (or channel username in the format @channelusername) */
  fromChatId: number | string;
  /** Message identifier in the chat specified in fromChatId */
  messageId: number;
  /** Sends the message silently. Users will receive a notification with no sound. */
  disableNotification?: boolean;
};

/**
 * Forward messages of any kind
 * @category Component
 * @props {@link ForwardMessageProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#forwardmessage).
 */
export const ForwardMessage: TelegramComponent<
  ForwardMessageProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function ForwardMessage(node, path) {
  const { fromChatId, disableNotification, messageId } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'forwardMessage',
      params: {
        from_chat_id: fromChatId,
        message_id: messageId,
        disable_notification: disableNotification,
      },
    }),
  ];
});

/**
 * @category Props
 */
export type ChatActionProps = {
  /** Type of action to broadcast depending on what the user is about to receive. */
  action:
    | 'typing'
    | 'upload_photo'
    | 'record_video'
    | 'upload_video'
    | 'record_audio'
    | 'upload_audio'
    | 'upload_document'
    | 'find_location'
    | 'record_video_note'
    | 'upload_video_note';
};

/**
 * Inform user that something is happening on the bot's side. The status is set for 5 seconds or less (when a message arrives from your bot, Telegram clients clear its typing status).
 * @category Component
 * @props {@link ChatActionProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#sendchataction).
 */
export const ChatAction: TelegramComponent<
  ChatActionProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function ChatAction(node, path) {
  const { action } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'sendChatAction',
      params: { action },
    }),
  ];
});

/**
 * @category Props
 */
export type KickChatMemberProps = {
  /** Unique identifier of the target user */
  userId: number;
  /** Date when the user will be unbanned, unix time. If user is banned for more than 366 days or less than 30 seconds from the current time they are considered to be banned forever */
  untilDate?: number | Date;
};

/**
 * Kick a user from a group, a supergroup or a channel. In the case of supergroups and channels, the user will not be able to return to the group on their own using invite links, etc., unless unbanned first. The bot must be an administrator in the chat for this to work and must have the appropriate admin rights.
 * @category Component
 * @props {@link KickChatMemberProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#kickchatmember).
 */
export const KickChatMember: TelegramComponent<
  KickChatMemberProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function KickChatMember(node, path) {
  const { userId, untilDate } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'kickChatMember',
      params: {
        user_id: userId,
        until_date:
          untilDate instanceof Date
            ? Math.round(untilDate.getTime() / 1000)
            : untilDate,
      },
    }),
  ];
});

/**
 * @category Props
 */
export type UnbanChatMemberProps = {
  /** Unique identifier of the target user */
  userId: number;
};

/**
 * Unban a previously kicked user in a supergroup or channel. The user will not return to the group or channel automatically, but will be able to join via link, etc. The bot must be an administrator for this to work.
 * @category Component
 * @props {@link UnbanChatMemberProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#unbanchatmember).
 */
export const UnbanChatMember: TelegramComponent<
  UnbanChatMemberProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function UnbanChatMember(node, path) {
  const { userId } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'unbanChatMember',
      params: {
        user_id: userId,
      },
    }),
  ];
});

/**
 * @category Props
 */
export type ChatPromotionProps = {
  /** True, if the user is allowed to send text messages, contacts, locations and venues */
  canSendMessages?: boolean;
  /** True, if the user is allowed to send audios, documents, photos, videos, video notes and voice notes, implies canSendMessages */
  canSendMediaMessages?: boolean;
  /** True, if the user is allowed to send polls, implies canSendMessages */
  canSendPolls?: boolean;
  /** True, if the user is allowed to send animations, games, stickers and use inline bots, implies canSendMediaMessages */
  canSendOtherMessages?: boolean;
  /** True, if the user is allowed to add web page previews to their messages, implies canSendMediaMessages */
  canAddWebPagePreviews?: boolean;
  /** True, if the user is allowed to change the chat title, photo and other settings. Ignored in public supergroups */
  canChangeInfo?: boolean;
  /** True, if the user is allowed to invite new users to the chat */
  canInviteUsers?: boolean;
  /** True, if the user is allowed to pin messages. Ignored in public supergroups */
  canPinMessages?: boolean;
  /** Date when restrictions will be lifted for the user, unix time. If user is restricted for more than 366 days or less than 30 seconds from the current time, they are considered to be restricted forever */
  untilDate?: number | Date;
};

/**
 * @category Props
 */
export type RestrictChatMemberProps = {
  /** Unique identifier of the target user */
  userId: number;
} & ChatPromotionProps;

/**
 * Restrict a user in a supergroup. The bot must be an administrator in the supergroup for this to work and must have the appropriate admin rights. Pass True for all permissions to lift restrictions from a user.
 * @category Component
 * @props {@link RestrictChatMemberProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#restrictchatmember).
 */
export const RestrictChatMember: TelegramComponent<
  RestrictChatMemberProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function RestrictChatMember(node, path) {
  const {
    userId,
    canSendMessages,
    canSendMediaMessages,
    canSendPolls,
    canSendOtherMessages,
    canAddWebPagePreviews,
    canChangeInfo,
    canInviteUsers,
    canPinMessages,
    untilDate,
  } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'restrictChatMember',
      params: {
        user_id: userId,
        permisions: {
          can_send_messages: canSendMessages,
          can_send_media_messages: canSendMediaMessages,
          can_send_polls: canSendPolls,
          can_send_other_messages: canSendOtherMessages,
          can_add_web_page_previews: canAddWebPagePreviews,
          can_change_info: canChangeInfo,
          can_invite_users: canInviteUsers,
          can_pin_messages: canPinMessages,
        },
        until_date:
          untilDate instanceof Date
            ? Math.round(untilDate.getTime() / 1000)
            : untilDate,
      },
    }),
  ];
});

/**
 * @category Props
 */
export type PromoteChatMemberProps = {
  /** Unique identifier of the target user */
  userId: number;
  /** Pass True, if the administrator can change chat title, photo and other settings */
  canChangeInfo?: boolean;
  /** Pass True, if the administrator can create channel posts, channels only */
  canPostMessages?: boolean;
  /** Pass True, if the administrator can edit messages of other users and can pin messages, channels only */
  canEditMessages?: boolean;
  /** Pass True, if the administrator can delete messages of other users */
  canDeleteMessages?: boolean;
  /** Pass True, if the administrator can invite new users to the chat */
  canInviteUsers?: boolean;
  /** Pass True, if the administrator can restrict, ban or unban chat members */
  canRestrictMembers?: boolean;
  /** Pass True, if the administrator can pin messages, supergroups only */
  canPinMessages?: boolean;
  /** Pass True, if the administrator can add new administrators with a subset of their own privileges or demote administrators that he has promoted, directly or indirectly (promoted by administrators that were appointed by him) */
  canPromoteMembers?: boolean;
};

/**
 * Promote or demote a user in a supergroup or a channel. The bot must be an administrator in the chat for this to work and must have the appropriate admin rights. Pass False for all boolean params to demote a user.
 * @category Component
 * @props {@link PromoteChatMemberProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#promotechatmember).
 */
export const PromoteChatMember: TelegramComponent<
  PromoteChatMemberProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function PromoteChatMember(node, path) {
  const {
    userId,
    canPostMessages,
    canEditMessages,
    canDeleteMessages,
    canRestrictMembers,
    canPromoteMembers,
    canChangeInfo,
    canInviteUsers,
    canPinMessages,
  } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'promoteChatMember',
      params: {
        user_id: userId,
        can_post_messages: canPostMessages,
        can_edit_messages: canEditMessages,
        can_delete_messages: canDeleteMessages,
        can_restrict_members: canRestrictMembers,
        can_promote_members: canPromoteMembers,
        can_change_info: canChangeInfo,
        can_invite_users: canInviteUsers,
        can_pin_messages: canPinMessages,
      },
    }),
  ];
});

/**
 * @category Props
 */
export type SetChatAdministratorCustomTitleProps = {
  /** Unique identifier of the target user */
  userId: number;
  /** New custom title for the administrator; 0-16 characters, emoji are not allowed */
  customTitle: string;
};

/**
 * Set a custom title for an administrator in a supergroup promoted by the bot.
 * @category Component
 * @props {@link SetChatAdministratorCustomTitleProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#setchatadministratorcustomtitle).
 */
export const SetChatAdministratorCustomTitle: TelegramComponent<
  SetChatAdministratorCustomTitleProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function SetChatAdministratorCustomTitle(node, path) {
  const { userId, customTitle } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'setChatAdministratorCustomTitle',
      params: {
        user_id: userId,
        custom_title: customTitle,
      },
    }),
  ];
});

/**
 * Set default chat permissions for all members. The bot must be an administrator in the group or a supergroup for this to work and must have the can_restrict_members admin rights.
 * @category Component
 * @props {@link ChatPromotionProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#setchatpermissions).
 */
export const SetChatPermissions: TelegramComponent<
  ChatPromotionProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function SetChatPermissions(node, path) {
  const {
    canSendMessages,
    canSendMediaMessages,
    canSendPolls,
    canSendOtherMessages,
    canAddWebPagePreviews,
    canChangeInfo,
    canInviteUsers,
    canPinMessages,
  } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'setChatPermissions',
      params: {
        permisions: {
          can_send_messages: canSendMessages,
          can_send_media_messages: canSendMediaMessages,
          can_send_polls: canSendPolls,
          can_send_other_messages: canSendOtherMessages,
          can_add_web_page_previews: canAddWebPagePreviews,
          can_change_info: canChangeInfo,
          can_invite_users: canInviteUsers,
          can_pin_messages: canPinMessages,
        },
      },
    }),
  ];
});

/**
 * @category Props
 */
export type SetChatPhotoProps = {
  /** The file content data. */
  fileData: Buffer | NodeJS.ReadableStream;
  /** Metadata about the uploading `fileData` if needed (while using Buffer). */
  fileInfo?: UploadingFileInfo;
};

/**
 * Set a new profile photo for the chat. Photos can't be changed for private chats. The bot must be an administrator in the chat for this to work and must have the appropriate admin rights.
 * @category Component
 * @props {@link SetChatPhotoProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#setchatphoto).
 */
export const SetChatPhoto: TelegramComponent<
  SetChatPhotoProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function SetChatPhoto(node, path) {
  const { fileData, fileInfo } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'setChatPhoto',
      params: {
        photo: undefined,
      },
      uploadingFiles: [
        { fieldName: 'photo', fileData, fileInfo, assetTag: undefined },
      ],
    }),
  ];
});

/**
 * Delete a chat photo. Photos can't be changed for private chats. The bot must be an administrator in the chat for this to work and must have the appropriate admin rights.
 * @category Component
 * @props `{}`
 * @guides Check official [reference](https://core.telegram.org/bots/api#deletechatphoto).
 */
export const DeleteChatPhoto: TelegramComponent<
  {},
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function DeleteChatPhoto(node, path) {
  return [
    makeUnitSegment(node, path, {
      method: 'deleteChatPhoto',
      params: {},
    }),
  ];
});

/**
 * @category Props
 */
export type SetChatTitleProps = {
  /** New chat title, 1-255 characters */
  title: string;
};

/**
 *  Change the title of a chat. Titles can't be changed for private chats. The bot must be an administrator in the chat for this to work and must have the appropriate admin rights.
 * @category Component
 * @props {@link SetChatTitleProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#setchattitle).
 */
export const SetChatTitle: TelegramComponent<
  SetChatTitleProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function SetChatTitle(node, path) {
  const { title } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'setChatTitle',
      params: { title },
    }),
  ];
});

/**
 * @category Props
 */
export type SetChatDescriptionProps = {
  /** New chat title, 1-255 characters */
  description: string;
};

/**
 * Change the description of a group, a supergroup or a channel. The bot must be an administrator in the chat for this to work and must have the appropriate admin rights.
 * @category Component
 * @props {@link SetChatDescriptionProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#setchatdescription).
 */
export const SetChatDescription: TelegramComponent<
  SetChatDescriptionProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function SetChatDescription(node, path) {
  const { description } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'setChatDescription',
      params: { description },
    }),
  ];
});

/**
 * @category Props
 */
export type PinChatMessageProps = {
  /** Identifier of a message to pin */
  messageId: number;
  /** Pass True, if it is not necessary to send a notification to all chat members about the new pinned message. Notifications are always disabled in channels. */
  disableNotification?: boolean;
};

/**
 * Pin a message in a group, a supergroup, or a channel. The bot must be an administrator in the chat for this to work and must have the 'can_pin_messages' admin right in the supergroup or 'can_edit_messages' admin right in the channel.
 * @category Component
 * @props {@link PinChatMessageProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#pinchatmessage).
 */
export const PinChatMessage: TelegramComponent<
  PinChatMessageProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function PinChatMessage(node, path) {
  const { messageId, disableNotification } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'pinChatMessage',
      params: {
        message_id: messageId,
        disable_notification: disableNotification,
      },
    }),
  ];
});

/**
 * Unpin a message in a group, a supergroup, or a channel. The bot must be an administrator in the chat for this to work and must have the 'can_pin_messages' admin right in the supergroup or 'can_edit_messages' admin right in the channel.
 * @category Component
 * @props `{}`
 * @guides Check official [reference](https://core.telegram.org/bots/api#unpinchatmessage).
 */
export const UnpinChatMessage: TelegramComponent<
  {},
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function UnpinChatMessage(node, path) {
  return [
    makeUnitSegment(node, path, {
      method: 'unpinChatMessage',
      params: {},
    }),
  ];
});

/**
 * Leave a group, supergroup or channel.
 * @category Component
 * @props `{}`
 * @guides Check official [reference](https://core.telegram.org/bots/api#leavechat).
 */
export const LeaveChat: TelegramComponent<
  {},
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function LeaveChat(node, path) {
  return [
    makeUnitSegment(node, path, {
      method: 'leaveChat',
      params: {},
    }),
  ];
});

/**
 * @category Props
 */
export type SetChatStickerSetProps = {
  /** Name of the sticker set to be set as the group sticker set */
  stickerSetName: string;
};

/**
 * Change the description of a group, a supergroup or a channel. The bot must be an administrator in the chat for this to work and must have the appropriate admin rights.
 * @category Component
 * @props {@link SetChatStickerSetProps}
 * @guides Check official [reference](https://core.telegram.org/bots/api#setchatstickerset).
 */
export const SetChatStickerSet: TelegramComponent<
  SetChatStickerSetProps,
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function SetChatStickerSet(node, path) {
  const { stickerSetName } = node.props;

  return [
    makeUnitSegment(node, path, {
      method: 'setChatStickerSet',
      params: { sticker_set_name: stickerSetName },
    }),
  ];
});

/**
 * Leave a group, supergroup or channel.
 * @category Component
 * @props `{}`
 * @guides Check official [reference](https://core.telegram.org/bots/api#deletechatstickerset).
 */
export const DeleteChatStickerSet: TelegramComponent<
  {},
  UnitSegment<TelegramSegmentValue>
> = makeTelegramComponent(function DeleteChatStickerSet(node, path) {
  return [
    makeUnitSegment(node, path, {
      method: 'deleteChatStickerSet',
      params: {},
    }),
  ];
});
