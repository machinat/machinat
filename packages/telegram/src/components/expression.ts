import { SociablyNode } from '@sociably/core';
import { makeUnitSegment, IntermediateSegment } from '@sociably/core/renderer';
import makeTelegramComponent from '../utils/makeTelegramComponent.js';
import {
  TelegramComponent,
  TelegramSegmentValue,
  TelegramParseMode,
} from '../types.js';

/** @category Props */
export type ExpressionProps = {
  children: SociablyNode;
  disableNotification?: boolean;
  parseMode?: TelegramParseMode;
  replyMarkup?: SociablyNode;
};

/**
 * Control options including disableNotification, parseMode of a group of
 * messages. Or add a replyMarkup at the last message in the Expression.
 *
 * @category Component
 * @props {@link ExpressionProps}
 */
export const Expression: TelegramComponent<
  ExpressionProps,
  IntermediateSegment<TelegramSegmentValue>
> = makeTelegramComponent(async function Expression(node, path, render) {
  const {
    children,
    disableNotification,
    parseMode = 'HTML',
    replyMarkup,
  } = node.props;

  const contentSegments: null | IntermediateSegment<TelegramSegmentValue>[] =
    await render(children, '.children');

  if (contentSegments === null) {
    return null;
  }

  const replyMarkupSegments = await render(replyMarkup, '.replyMarkup');

  const outputSegments: IntermediateSegment<TelegramSegmentValue>[] = [];
  let lastEmptyMarkupSlot = -1;

  contentSegments.forEach((segment) => {
    if (segment.type === 'text') {
      lastEmptyMarkupSlot = outputSegments.length;

      outputSegments.push(
        makeUnitSegment(node, path, {
          method: 'sendMessage',
          params: {
            text: segment.value,
            parse_mode: parseMode === 'None' ? undefined : parseMode,
            disable_notification: disableNotification,
          },
        }),
      );
    } else if (segment.type === 'unit' || segment.type === 'raw') {
      const { method } = segment.value;

      if (method.startsWith('send') && method !== 'sendChatAction') {
        const { params } = segment.value;
        if (!params.reply_markup) {
          lastEmptyMarkupSlot = outputSegments.length;
        }

        outputSegments.push({
          ...segment,
          value: {
            ...segment.value,
            params: {
              ...params,
              disable_notification:
                typeof params.disable_notification === 'undefined'
                  ? disableNotification
                  : params.disable_notification,
            },
          },
        });
      } else {
        outputSegments.push(segment);
      }
    } else if (segment.type !== 'break') {
      outputSegments.push(segment);
    }
  });

  if (replyMarkupSegments !== null) {
    if (lastEmptyMarkupSlot === -1) {
      throw new Error('no message available to attach reply markup');
    }
    outputSegments[lastEmptyMarkupSlot].value.params.reply_markup =
      replyMarkupSegments[0].value;
  }

  return outputSegments;
});
