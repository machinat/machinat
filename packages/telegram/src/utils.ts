import { makeNativeComponent } from '@sociably/core/renderer';
import { TELEGRAM } from './constant';

// eslint-disable-next-line import/prefer-default-export
export const makeTelegramComponent = makeNativeComponent(TELEGRAM);
