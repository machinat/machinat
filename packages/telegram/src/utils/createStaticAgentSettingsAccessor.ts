import type TelegramUser from '../User';
import { AgentSettingsAccessorI } from '../interface';
import { TelegramAgentSettings } from '../types';

const createStaticAgentSettingsAccessor = (
  settings: TelegramAgentSettings[]
): AgentSettingsAccessorI => {
  const findSettingsWithMatchedToken = ({ id: botId }: TelegramUser) =>
    settings.find(({ botToken }) => botToken.startsWith(`${botId}:`)) || null;

  return {
    getAgentSettings: async (bot) => findSettingsWithMatchedToken(bot),
    getAgentSettingsBatch: async (bots) =>
      bots.map(findSettingsWithMatchedToken),
  };
};

export default createStaticAgentSettingsAccessor;
