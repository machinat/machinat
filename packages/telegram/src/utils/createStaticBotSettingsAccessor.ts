import type TelegramUser from '../User';
import { BotSettingsAccessorI } from '../interface';
import { TelegramBotSettings } from '../types';

const createStaticBotSettingsAccessor = (
  settings: TelegramBotSettings[]
): BotSettingsAccessorI => {
  const findSettingsWithMatchedToken = ({ id: botId }: TelegramUser) =>
    settings.find(({ botToken }) => botToken.startsWith(`${botId}:`)) || null;

  return {
    getChannelSettings: async (bot) => findSettingsWithMatchedToken(bot),
    getChannelSettingsBatch: async (bots) =>
      bots.map(findSettingsWithMatchedToken),
    listAllChannelSettings: async () => settings,
  };
};

export default createStaticBotSettingsAccessor;
