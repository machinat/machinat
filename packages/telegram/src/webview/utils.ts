import type { ContextDetails } from '@sociably/auth';
import TelegramChat from '../Chat';
import TelegramUser from '../User';
import type { RawUser } from '../types';
import type { TelegramAuthContext, TelegramAuthData } from './types';

// eslint-disable-next-line import/prefer-default-export
export const getAuthContextDetails = ({
  chat: chatData,
  botId,
  botName,
  user: userData,
  photo: photoUrl,
}: TelegramAuthData): ContextDetails<TelegramAuthContext> => {
  return {
    channel: new TelegramUser(botId, true),
    user: new TelegramUser(
      userData.id,
      false,
      { ...userData, is_bot: false },
      photoUrl
    ),
    thread: chatData
      ? new TelegramChat(botId, chatData.id, chatData)
      : TelegramChat.fromUser(botId, userData as RawUser),
    botId,
    botName,
    photoUrl,
  };
};
