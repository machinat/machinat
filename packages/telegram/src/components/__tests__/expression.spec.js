import Machinat from '@machinat/core';
import { isNativeType } from '@machinat/core/utils';
import Renderer from '@machinat/core/renderer';
import renderGeneralType from '../general';
import { ChatAction } from '../action';
import { Photo } from '../media';
import { ForceReply } from '../replyMarkup';
import { Expression } from '../expression';

const renderer = new Renderer('telegram', renderGeneralType);

it('is valid unit Component', () => {
  expect(typeof Expression).toBe('function');
  expect(isNativeType(<Expression />)).toBe(true);
  expect(Expression.$$platform).toBe('telegram');
});

it('hoist text value into message object', async () => {
  const result = await renderer.render(
    <Expression>
      <p>foo</p>
      <p>
        <b>bar</b>
      </p>
      baz
    </Expression>
  );
  expect(result[0].value.method).toBe('sendMessage');
  expect(result[0].value.parameters.text).toBe('foo');
  expect(result[1].value.method).toBe('sendMessage');
  expect(result[1].value.parameters.text).toBe('<b>bar</b>');
  expect(result[2].value.method).toBe('sendMessage');
  expect(result[2].value.parameters.text).toBe('baz');
  expect().toMatchSnapshot();
});

it('set disableNotification in message parameters', async () => {
  const result = await renderer.render(
    <Expression disableNotification>
      foo
      <Photo url="http://machinat.com/bar.jpg" />
    </Expression>
  );
  expect(result[0].value.parameters.disable_notification).toBe(true);
  expect(result[1].value.parameters.disable_notification).toBe(true);
  expect().toMatchSnapshot();
});

it('respect the original disableNotification setting set on the messgage', async () => {
  const result = await renderer.render(
    <Expression disableNotification>
      <Photo url="http://machinat.com/bar.jpg" />
      <Photo disableNotification={false} url="http://machinat.com/baz.jpg" />
    </Expression>
  );
  expect(result[0].value.parameters.disable_notification).toBe(true);
  expect(result[1].value.parameters.disable_notification).toBe(false);
  expect(result).toMatchSnapshot();
});

it('add replyMarkup to the last supported message', async () => {
  let result = await renderer.render(
    <Expression disableNotification replyMarkup={<ForceReply />}>
      <Photo url="http://machinat.com/foo.jpg" />
      <ChatAction action="upload_photo" />
    </Expression>
  );
  expect(result).toMatchSnapshot();
  expect(result[0].value.parameters.reply_markup).toMatchInlineSnapshot(`
    Object {
      "force_reply": true,
      "selective": undefined,
    }
  `);

  result = await renderer.render(
    <Expression disableNotification replyMarkup={<ForceReply />}>
      foo
      <Photo url="http://machinat.com/foo.jpg" replyMarkup={<ForceReply />} />
    </Expression>
  );
  expect(result).toMatchSnapshot();
  expect(result[0].value.parameters.reply_markup).toMatchInlineSnapshot(`
    Object {
      "force_reply": true,
      "selective": undefined,
    }
  `);
});

it('throw if no message available to attach replyMarkup', async () => {
  await expect(
    renderer.render(
      <Expression disableNotification replyMarkup={<ForceReply />}>
        <Photo url="http://machinat.com/foo.jpg" replyMarkup={<ForceReply />} />
        <ChatAction action="upload_photo" />
      </Expression>
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"no message available to attach reply markup"`
  );

  await expect(
    renderer.render(
      <Expression disableNotification replyMarkup={<ForceReply />}>
        <ChatAction action="upload_photo" />
      </Expression>
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"no message available to attach reply markup"`
  );
});
