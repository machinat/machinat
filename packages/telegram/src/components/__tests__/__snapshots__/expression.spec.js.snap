// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add replyMarkup to the last supported message 1`] = `
Array [
  Object {
    "node": <Photo
      url="http://machinat.com/foo.jpg"
    />,
    "path": "$#Expression.children:0",
    "type": "unit",
    "value": Object {
      "method": "sendPhoto",
      "parameters": Object {
        "caption": undefined,
        "disable_notification": true,
        "parse_mode": "HTML",
        "photo": "http://machinat.com/foo.jpg",
        "reply_markup": Object {
          "force_reply": true,
          "selective": undefined,
        },
        "reply_to_message_id": undefined,
      },
      "uploadingFiles": undefined,
    },
  },
  Object {
    "node": <ChatAction
      action="upload_photo"
    />,
    "path": "$#Expression.children:1",
    "type": "unit",
    "value": Object {
      "method": "sendChatAction",
      "parameters": Object {
        "action": "upload_photo",
      },
    },
  },
]
`;

exports[`add replyMarkup to the last supported message 3`] = `
Array [
  Object {
    "node": <Expression
      disableNotification={true}
      replyMarkup={<ForceReply />}
    >
      foo
      <Photo
        replyMarkup={<ForceReply />}
        url="http://machinat.com/foo.jpg"
      />
    </Expression>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "method": "sendMessage",
      "parameters": Object {
        "disable_notification": true,
        "parse_mode": "HTML",
        "reply_markup": Object {
          "force_reply": true,
          "selective": undefined,
        },
        "text": "foo",
      },
    },
  },
  Object {
    "node": <Photo
      replyMarkup={<ForceReply />}
      url="http://machinat.com/foo.jpg"
    />,
    "path": "$#Expression.children:1",
    "type": "unit",
    "value": Object {
      "method": "sendPhoto",
      "parameters": Object {
        "caption": undefined,
        "disable_notification": true,
        "parse_mode": "HTML",
        "photo": "http://machinat.com/foo.jpg",
        "reply_markup": Object {
          "force_reply": true,
          "selective": undefined,
        },
        "reply_to_message_id": undefined,
      },
      "uploadingFiles": undefined,
    },
  },
]
`;

exports[`hoist text value into message object 1`] = `undefined`;

exports[`respect the original disableNotification setting set on the messgage 1`] = `
Array [
  Object {
    "node": <Photo
      url="http://machinat.com/bar.jpg"
    />,
    "path": "$#Expression.children:0",
    "type": "unit",
    "value": Object {
      "method": "sendPhoto",
      "parameters": Object {
        "caption": undefined,
        "disable_notification": true,
        "parse_mode": "HTML",
        "photo": "http://machinat.com/bar.jpg",
        "reply_markup": undefined,
        "reply_to_message_id": undefined,
      },
      "uploadingFiles": undefined,
    },
  },
  Object {
    "node": <Photo
      disableNotification={false}
      url="http://machinat.com/baz.jpg"
    />,
    "path": "$#Expression.children:1",
    "type": "unit",
    "value": Object {
      "method": "sendPhoto",
      "parameters": Object {
        "caption": undefined,
        "disable_notification": false,
        "parse_mode": "HTML",
        "photo": "http://machinat.com/baz.jpg",
        "reply_markup": undefined,
        "reply_to_message_id": undefined,
      },
      "uploadingFiles": undefined,
    },
  },
]
`;

exports[`set disableNotification in message parameters 1`] = `undefined`;
